src/ir/possible-contents.cpp

```diff
@@ -525,21 +524,53 @@ struct InfoCollector
         return SignatureResultLocation{targetType, i};
       });
   }
-  void visitCallRef(CallRef* curr) {
-    auto targetType = curr->target->type;
+  template<typename T> void handleIndirectCall(T* curr, Type targetType) {
+    // If the type is unreachable, nothing can be called (and there is no heap
+    // type to get).
     if (targetType != Type::unreachable) {
-      auto heapType = targetType.getHeapType();
-      handleCall(
-        curr,
-        [&](Index i) {
-          return SignatureParamLocation{heapType, i};
-        },
-        [&](Index i) {
-          return SignatureResultLocation{heapType, i};
-        });
+      handleIndirectCall(curr, targetType.getHeapType());
     }
   }
 
+  void visitCall(Call* curr) {
+    Name targetName;
+    if (!Intrinsics(*getModule()).isCallWithoutEffects(curr)) {
+      // This is just a normal call.
+      handleDirectCall(curr, curr->target);
+      return;
+    }
+    // A call-without-effects receives a function reference and calls it, the
+    // same as a CallRef. When we have a flag for non-closed-world, we should
+    // handle this automatically by the reference flowing out to an import,
+    // which is what binaryen intrinsics look like. For now, to support use
+    // cases of a closed world but that also use this intrinsic, handle the
+    // intrinsic specifically here.
```

Comment:
> When we have a flag for non-closed-world,

What is a flag for non-closed world?

> we should handle this automatically by the reference flowing out to an import,

What does "by the reference flowing out to an import"? Does this mean you call an imported function? Why is that "flowing out"?

> which is what binaryen intrinsics look like.

What does this mean? Aren't we handling binaryen intrinsics here..? Are there two kinds of intrinsics?

Reply:
> What is a flag for non-closed world?

Right now in GC we assume a closed world, which is what all producers assume anyhow. That assumption makes it ok in e.g. GlobalTypeOptimization to optimize a type (remove a field, say) even if that type "escapes" to the outside. If the module were linked with another, such optimizations could be noticeable, so they are technically unsafe. We do them for now, but eventually we'll want to have a flag to opt-in to a closed world assumption, so the default behavior is safe. That will take work to find which types "escape" to the outside.

> What does "by the reference flowing out to an import"? Does this mean you call an imported function? Why is that "flowing out"?

Sorry, that wasn't clear. I meant, if a `ref.func` can reach the outside of the module - by being returned from an export, for example - then the function might be called from the outside. Normally we assume that just seeing a `ref.func` doesn't mean the contents of that function are used - we also need to see a call that might operate on it.

In a closed world, which we assume for now, not seeing such a call means it cannot be called. The problem is that a binaryen intrinsic is exactly that, a `ref.func` is sent to the outside of the module (sent to an import), which does the call there. We need to be aware that the function is called somewhere.

> Are there two kinds of intrinsics?

Sorry, that wasn't clear either I think. This is all about binaryen intrinsics, which is just the `call.without.effects` intrinsic really.

Reply:
I think the key thing missing from the comment that would make it clear is an explanation of what goes wrong without this special handling.